src/Adapter/Discount/Update/DiscountDuplicator.php

```diff
@@ -0,0 +1,331 @@
+                ->setParameter('id_type', $typeId)
+                ->executeQuery();
+        }
+    }
+
+    /**
+     * Duplicates customer groups associated to a discount
+     */
+    private function duplicateGroups(DiscountId $oldDiscountId, DiscountId $newDiscountId): void
+    {
+        $groups = $this->discountRepository->getGroups($oldDiscountId);
+
+        foreach ($groups as $groupId) {
+            $qb = $this->connection->createQueryBuilder();
+            $qb->insert($this->dbPrefix . 'cart_rule_group')
+                ->values([
+                    'id_cart_rule' => ':id_discount',
+                    'id_group' => ':id_group',
+                ])
+                ->setParameter('id_discount', $newDiscountId->getValue())
+                ->setParameter('id_group', $groupId)
+                ->executeQuery();
+        }
+    }
+
+    /**
+     * Duplicates shops associated to a discount
+     */
+    private function duplicateShops(DiscountId $oldDiscountId, DiscountId $newDiscountId): void
+    {
+        $shops = $this->discountRepository->getShops($oldDiscountId);
+
+        foreach ($shops as $shopId) {
+            $qb = $this->connection->createQueryBuilder();
+            $qb->insert($this->dbPrefix . 'cart_rule_shop')
+                ->values([
+                    'id_cart_rule' => ':id_discount',
+                    'id_shop' => ':id_shop',
+                ])
+                ->setParameter('id_discount', $newDiscountId->getValue())
+                ->setParameter('id_shop', $shopId)
+                ->executeQuery();
+        }
+    }
+
+    /**
+     * Duplicates product rules associated to a discount
+     */
+    private function duplicateProductRules(DiscountId $oldDiscountId, DiscountId $newDiscountId): void
+    {
+        // First, we need to copy the rule groups (with their new IDs)
+        $ruleGroupsIds = [];
+        $qb = $this->connection->createQueryBuilder();
+        $ruleGroups = $qb
+            ->select('*')
+            ->from($this->dbPrefix . 'cart_rule_product_rule_group')
+            ->where('id_cart_rule = :id_discount')
+            ->setParameter('id_discount', $oldDiscountId->getValue())
+            ->executeQuery()
+            ->fetchAllAssociative();
+
+        foreach ($ruleGroups as $ruleGroup) {
+            $qb = $this->connection->createQueryBuilder();
+            $qb->insert($this->dbPrefix . 'cart_rule_product_rule_group')
+                ->values([
+                    'id_cart_rule' => ':id_discount',
+                    'quantity' => ':quantity',
+                    'type' => ':type',
+                ])
+                ->setParameter('id_discount', $newDiscountId->getValue())
+                ->setParameter('quantity', $ruleGroup['quantity'])
+                ->setParameter('type', $ruleGroup['type'])
+                ->executeQuery();
+            $ruleGroupsIds[$ruleGroup['id_product_rule_group']] = $this->connection->lastInsertId();
```

Comment:
Hmm be careful, you should create a new array for dedicated new rule groups IDs

```suggestion
            $newRuleGroupsIds[$ruleGroup['id_product_rule_group']] = $this->connection->lastInsertId();
```